File: src/main/java/io/mfedirko/admin/LearningAdminController.java
package io.mfedirko.admin;

import io.mfedirko.learning.CreateLessonForm;
import io.mfedirko.learning.LearningRepository;
import io.mfedirko.learning.Lesson;
import io.mfedirko.learning.UpdateLessonForm;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.validation.Errors;
import org.springframework.web.bind.annotation.*;

@Controller
@RequestMapping("/admin/learning")
@RequiredArgsConstructor
public class LearningAdminController {
    public static final String CREATE_LESSON = "admin/create-lesson";
    public static final String UPDATE_LESSON = "admin/update-lesson";
    public static final String LESSON_CARD = "fragments/lesson-card";
    private final LearningRepository repository;

    @GetMapping("/create-form")
    public String getCreateLessonForm(ModelMap modelMap) {
        modelMap.addAttribute("createLessonForm", CreateLessonForm.builder().build());
        return CREATE_LESSON;
    }

    @PostMapping("/create-form")
    public String submitCreateLesson(@Valid CreateLessonForm createLessonForm, Errors errors, ModelMap modelMap) {
        if (errors.hasErrors()) {
            return CREATE_LESSON;
        }
        long id = repository.createLesson(createLessonForm);
        return lessonCard(modelMap, id);
    }

    @GetMapping("/update-form/{id}")
    public String getUpdateLessonForm(@PathVariable("id") long id, ModelMap modelMap) {
        Lesson lesson = repository.getLesson(id);
        modelMap.addAttribute("updateLessonForm", UpdateLessonForm.fromLesson(lesson));
        return UPDATE_LESSON;
    }

    @PostMapping("/update-form/{id}")
    public String submitUpdateLesson(@PathVariable("id") long id, @Valid UpdateLessonForm form, Errors errors, ModelMap modelMap) {
        if (errors.hasErrors()) {
            return UPDATE_LESSON;
        }
        repository.updateLesson(form, id);
        return lessonCard(modelMap, id);
    }

    @GetMapping("/update-cancel/{id}")
    public String cancelUpdate(@PathVariable("id") long id, ModelMap modelMap) {
        return lessonCard(modelMap, id);
    }

    @DeleteMapping("/{id}")
    @ResponseBody
    public String deleteLesson(@PathVariable("id") long id) {
        repository.deleteLesson(id);
        return "Deleted " + id;
    }

    private String lessonCard(ModelMap modelMap, long id) {
        Lesson lesson = repository.getLesson(id);
        modelMap.addAttribute("lesson", lesson);
        return LESSON_CARD;
    }
}

File: src/main/kotlin/io/mfedirko/admin/LearningAdminController.kt
package io.mfedirko.admin

import io.mfedirko.learning.CreateLessonForm
import io.mfedirko.learning.LearningRepository
import io.mfedirko.learning.UpdateLessonForm
import io.mfedirko.learning.UpdateLessonForm.Companion.fromLesson
import jakarta.validation.Valid
import org.springframework.stereotype.Controller
import org.springframework.ui.ModelMap
import org.springframework.validation.Errors
import org.springframework.web.bind.annotation.*

@Controller
@RequestMapping("/admin/learning")
class LearningAdminController(
    private val repository: LearningRepository
) {
    @GetMapping("/create-form")
    fun getCreateLessonForm(modelMap: ModelMap): String {
        modelMap.addAttribute("createLessonForm", CreateLessonForm())
        return CREATE_LESSON
    }

    @PostMapping("/create-form")
    fun submitCreateLesson(@Valid createLessonForm: CreateLessonForm, errors: Errors, modelMap: ModelMap): String {
        if (errors.hasErrors()) {
            return CREATE_LESSON
        }
        val id = repository.createLesson(createLessonForm)
        return lessonCard(modelMap, id)
    }

    @GetMapping("/update-form/{id}")
    fun getUpdateLessonForm(@PathVariable("id") id: Long, modelMap: ModelMap): String {
        val lesson = repository.getLesson(id)!!
        modelMap.addAttribute("updateLessonForm", fromLesson(lesson))
        return UPDATE_LESSON
    }

    @PostMapping("/update-form/{id}")
    fun submitUpdateLesson(
        @PathVariable("id") id: Long,
        @Valid form: UpdateLessonForm,
        errors: Errors,
        modelMap: ModelMap
    ): String {
        if (errors.hasErrors()) {
            return UPDATE_LESSON
        }
        repository.updateLesson(form, id)
        return lessonCard(modelMap, id)
    }

    @GetMapping("/update-cancel/{id}")
    fun cancelUpdate(@PathVariable("id") id: Long, modelMap: ModelMap): String {
        return lessonCard(modelMap, id)
    }

    @DeleteMapping("/{id}")
    @ResponseBody
    fun deleteLesson(@PathVariable("id") id: Long): String {
        repository.deleteLesson(id)
        return "Deleted $id"
    }

    private fun lessonCard(modelMap: ModelMap, id: Long): String {
        val lesson = repository.getLesson(id)
        modelMap.addAttribute("lesson", lesson)
        return LESSON_CARD
    }

    companion object {
        const val CREATE_LESSON = "admin/create-lesson"
        const val UPDATE_LESSON = "admin/update-lesson"
        const val LESSON_CARD = "fragments/lesson-card"
    }
}